### ISSUE TITLE
'.git/refs' is insufficient as generateAppVersionInfo task input

### ISSUE THREAD

**Issue by Wrywulf** [ReactiveCircus/app-versioning#27]
We ran into an interesting problem with our CI builds.

Our apk version name/code is parsed from git tags. When building "play store" apks our CI deployment task checks out a git tag and builds directly from that.

However, we noticed that some of our "play store" apks did not generate the correct version name/code from the tag. The generateAppVersionInfo task was not executed, it was either UP-TO-DATE or FROM-CACHE.

I tracked it down to the git refs task input property

```
    @get:Optional
    @get:InputDirectory
    @get:PathSensitive(PathSensitivity.RELATIVE)
    @get:IgnoreEmptyDirectories
    @get:NormalizeLineEndings
    abstract val gitRefsDirectory: DirectoryProperty
```

When simply changing the git HEAD (which is what our CI deployment task is doing by checking out an existing commit) the `.git/refs` dir is not changing. Thus, the task output is not considered stale - even though it should be.

Afaict the tests for the plugin only checks the incremental and cacheable properties of the plugin task when *creating a new commit/tag*:
https://github.com/ReactiveCircus/app-versioning/blob/main/src/functionalTest/kotlin/io/github/reactivecircus/appversioning/tasks/GenerateAppVersionInfoTest.kt

I'm not sure what the proper solution is here. But I suppose it would make sense to make the `.git/HEAD` an input for the generateAppVersionInfo task as well?
Afaict that would make its output stale when switching between existing commits/tags.

**Comment by ychescale9** [comment]
Thanks for raising this with great analysis, I'll take a closer look next week.

**Comment by ychescale9** [comment]
@Wrywulf just wanted to clarify your use case: when you change git HEAD on CI, are you checking out a commit with an existing tag which is not the latest tag?

**Comment by ychescale9** [comment]
I took a closer look and did some tests locally. When checking out different commits with tags, the tasks become up-to-date only after the first run. As soon as the tag is changed the task will be re-executed.

I still think git tag should be the only external input that can invalidate the cache but maybe I'm missing something with your use case.

**Comment by Wrywulf** [comment]
Thanks for taking a look.

Yeah, on CI our use case is:

1. clone git repo (entirely)
2. checkout commit (headless) with `id`
3. run build

The `.git/refs` folder is not affected by step (2), which also prevents the `generateAppVersionInfo` from being marked as stale. Our CI agents use both local and remote gradle caches.

I tried to verify the above by running
```
ls -alRc .git/refs | sha1sum
```
between changes to the git repo. Adding new commits or tags changes the sha1. But simply moving the git head does not (which makes sence since the HEAD is tracked at `.git/HEAD`.

Does that information help?

**Comment by ychescale9** [comment]
In step 2 does the commit have a tag? When do you add a new tag for the release?

**Comment by Wrywulf** [comment]
Yeah, the commit `id` already has a tag at step (2) - unless something else went wrong.

The commit `id` is the "primary" artifact that we transfer from our build pipeline -> deployment pipeline. So most of our deployment steps use it. The first deployment steps tags the commit `id` with our `x.y.z` release tag. But the following steps still use the commit `id` as reference.

**Comment by ychescale9** [comment]
>Each step can run on a different CI agent (and that's really where the caching probs kick in)

when this happens what's the generated version? is it an an version generated by an old tag?

**Comment by Wrywulf** [comment]
I mean, there are other conditions that need to fall into place to trigger the scenario. But all other things being equal, yes, its returns the locally cached generated version (from a previous build where `.git/refs` was identical)

**Comment by ychescale9** [comment]
ok I think I understand your scenario. I'm not sure how / if the plugin should handle this though. If given the same tag 2 builds are expected to have different versions (the task re-run), the variables that can affect the generated version need to be part of the cache input which the task doesn't know about.

As an alternative you can run `./gradlew generateAppVersionInfoFor<BuildVariant> --rerun-tasks` before the build. But I wonder if this will create the opposite issue for you where some builds are expected to have identical version?

**Comment by Wrywulf** [comment]
I think that perhaps we're not completely aligned on the problem I face.
I'll try to explain myself better 😅 

Consider a git repo with 2 tags at different commits: `1.0` and `1.1`.

Then consider the following work in sequence:

1. git clone `repo`
2. git checkout `1.0`
3. ./gradlew generateVersionInfoFor<variant>

Task is executed and version "1.0" is generated. (as expected)

4.  git checkout `1.1`
5. ./gradlew generateVersionInfoFor<variant>

Task is UP-TO-DATE and version is stil "1.0". This is because `.git/refs` didn't change when checking out `1.1` (only `.git/HEAD` changed).
The expected version is "1.1".

6. git checkout -b "temp" 
7. ./gradlew generateVersionInfoFor<variant>

Task is executed and version "1.1" is generated. (as expected)
By creating the temporary local branch `temp` I forced the `.git/refs` to change and thus invalidate the task input.


Does that make it clearer?

**Comment by ychescale9** [comment]
Thanks! I was able to reproduce the issue with a couple of tests:

https://github.com/ReactiveCircus/app-versioning/blob/9f8ea2ce70c1b9c90a062ffc79e1cfedee07567e/src/functionalTest/kotlin/io/github/reactivecircus/appversioning/tasks/GenerateAppVersionInfoTest.kt#L957-L1031

Will push a new release soon.

**Comment by ychescale9** [comment]
The fix is available in `1.1.2`. Let me know if that fixes it for you 😃

**Comment by Wrywulf** [comment]
Thanks! I'll give it a spin
